src/v/ssx/when_all.h

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include <seastar/core/future.hh>
+#include <seastar/coroutine/as_future.hh>
+
+#include <concepts>
+#include <ranges>
+#include <type_traits>
+
+namespace ssx {
+
+namespace detail {
+
+template<typename Container>
+concept emplace_backable = requires(Container c) {
+    typename Container::value_type;
+    c.emplace_back(std::declval<typename Container::value_type>());
+};
+
+} // namespace detail
+
+/// \brief Wait for a range of futures to complete.
+///
+/// Given a range of futures as input, wait for all of them
+/// to resolve, and return a future containing a range with the
+/// resolved values of the original futures.
+///
+/// If any future fails, one of the exceptions will be
+/// returned as a failed future.
+///
+/// \param futures a \c FutureRange containing the futures to wait for.
+/// \return a \c future<ResolvedContainer> with all the resolved values
+/// of \c futures
+template<typename ResolvedContainer, typename FutureRange>
+requires seastar::Future<typename FutureRange::value_type>
+         && std::ranges::input_range<FutureRange>
+         && std::ranges::
+           output_range<FutureRange, typename FutureRange::value_type>
+         && std::constructible_from<
+           typename ResolvedContainer::value_type,
+           decltype(std::declval<FutureRange>().begin()->get())>
+         && detail::emplace_backable<ResolvedContainer>
+inline seastar::future<ResolvedContainer>
+when_all_succeed(FutureRange futures) {
+    for (auto& fut : futures) {
+        fut = co_await seastar::coroutine::as_future(std::move(fut));
+    }
+
+    ResolvedContainer result{};
```

Comment:
You'll want to `reserve()` here to allow move-only types.

Reply:
Reserve is a good idea, but that doesn't enable move-only types right? The type system can't ensure that you've already pre-reserved space

Reply:
> Reserve is a good idea, but that doesn't enable move-only types right? The type system can't ensure that you've already pre-reserved space

Correct. I was thinking there might be something around nothrow(false) moveable, but that's always a bad idea, anyway.

Reply:
I agree. With vector, when you are resizing, you are only moving the elements, so it should work without the `reserve`. Still a good thing to add, though.